numpy/_core/src/multiarray/nditer_constr.c

```diff
@@ -1931,6 +1923,390 @@ operand_different_than_broadcast: {
     }
 }
 
+
+/*
+ * At this point we (presumably) use a buffered iterator and here we want
+ * to find out the best way to buffer the iterator in a fashion that we don't
+ * have to figure out a lot of things on every outer iteration.
+ *
+ * How do we iterate?
+ * ------------------
+ * There are currently two modes of "buffered" iteration:
+ * 1. The normal mode, where we either buffer each operand or not and
+ *    then do a 1-D loop on those buffers (or operands).
+ * 2. The "reduce" mode.  In reduce mode (ITFLAG_REDUCE) we internally use a
+ *    a double iteration where for "reduce" operands we have:
+ *      - One outer iteration with stride == 0 and a core with at least one
+ *        stride != 0 (all of them if this is a true reduce/writeable operand).
+ *      - One outer iteration with stride != 0 and a core of all strides == 0.
+ *    This setup allows filling the buffer with only the stride != 0 and then
+ *    doing the double loop.
+ *
+ * Only a writeable (reduce) operand require this reduce mode because for
+ * reading it is OK if the buffer holds duplicates elements.
+ * The goal of the reduce mode is that it allows for larger core sizes and
+ * buffers since the zero strides do not allow a single 1-d iteration.
+ * If do we use the reduce-mode, we can apply it also to read-only operands
+ * as an optimization.
+ *
+ * The functino here finds an outer dimension and it's "core" to use that
+ * works with reductions.
+ * While iterating, we will fill the buffers making sure that we:
+ *   - Never buffer beyond the outer dimension (optimize chance of re-use).
```

Comment:
Added two examples, hopefully they help (if not for this, then hopefully to show how to think about these iterations).

The outer dimension may be iterated in chunks (not fully), so if it is say 100, and we can fit 30 of them in at a time, you reach 90.  Then we just buffer 10, not 30 again, because that would increment the dimension that comes next (with a different stride).